### DISCUSSION TITLE
TestStore & long running effect

### DISCUSSION THREAD

**Discussion by gkaimakas** [pointfreeco/swift-composable-architecture#3454] [Q&A]
Hey good people!

I have a `Reducer` that starts a long running effect when the view appears (with SwiftUI's .task modifier)

```swift
...
case .task:
let effA = .run { send in /// runs once
      await .send(.fetchObjsCompleted(Result(catching: {
         /// fetch some data
      }))
   }

let effB = .run { send in /// is long living
      for await _ in objCreatedAsyncStream {
        await .send(.fetchObjsCompleted(Result(catching: {
          /// fetch some data
        }))
      }
   }

return .concatenate(
   effA,
   effB
)
...

```

The rationale behind all that is that I want to fetch an initial batch of data, but also to subscribe to notifications from `objCreatedAsyncStream` and when an event is emitted, refresh the data.

I was able to test the initial fetching of data using `TestStore` but when I try to test the long running effect I'm facing issues. I want to return an empty array of objs from `effA` and once  `objCreatedAsyncStream` emits an event, return an array with some objects.

```swift

let store = TestStore(initialState: State()) {
    MyReducer()
} withDependencies: {
    $0.fetchData = { [] }
} 

await store.send(.task)
await store.receive(\.fetchObjsCompleted.success) {
    /// check that effA returns an empty array
}

/// effB is still running at this point

store.dependencies.fetchData = { [objA] }

objCreatedAsyncStream.send(value: objA) // a new event is emitted

await store.receive(\.fetchObjsCompleted.success) {
    /// check that effB return [objA]
}
```

I thought that I could swap the dependency that fetches the data during the test, but it seems that swapping a dependency with a long running effect, is not supported.

Any ideas on how to approach this?

**Comment by pyrtsa** [comment]
I'd introduce a local variable `let data = LockIsolated<[MyObject]>([])` next to the `TestStore` and, instead of ever reassigning `store.dependencies.fetchData`, simply initialise it to `$0.fetchData = { data.value }` and change `data` between assertions.

**Reply by gkaimakas** [reply]
That worked like a charm! Thanks @pyrtsa
